app/uk/gov/hmrc/eoricommoncomponent/frontend/services/cache/sessionCache.scala

```diff
@@ -106,120 +65,153 @@ object CachedData {
+  def saveEori(eori: Eori)(implicit request: Request[_]): Future[Boolean] =
+    putData(eoriKey, Json.toJson(eori.id)) map (_ => true)
 
-  def saveGroupEnrolment(groupEnrolment: EnrolmentResponse)(implicit hc: HeaderCarrier): Future[Boolean] =
-    createOrUpdate(sessionId, groupEnrolmentKey, Json.toJson(groupEnrolment)) map (_ => true)
+  def keepAlive(implicit request: Request[_]): Future[Boolean] =
+    putData(keepAliveKey, Json.toJson(LocalDateTime.now().toString)) map (_ => true)
 
-  private def getCached[T](sessionId: Id, t: (CachedData, Id) => T): Future[T] =
-    findById(sessionId.id).map {
-      case Some(Cache(_, Some(data), _, _)) =>
-        Json.fromJson[CachedData](data) match {
-          case d: JsSuccess[CachedData] => t(d.value, sessionId)
-          case _ =>
-            eccLogger.error(s"No Session data is cached for the sessionId : ${sessionId.id}")
-            throw SessionTimeOutException(s"No Session data is cached for the sessionId : ${sessionId.id}")
-        }
-      case _ =>
-        eccLogger.info(s"No match session id for signed in user with session: ${sessionId.id}")
-        throw SessionTimeOutException(s"No match session id for signed in user with session : ${sessionId.id}")
-    }
+  def saveGroupEnrolment(groupEnrolment: EnrolmentResponse)(implicit request: Request[_]): Future[Boolean] =
+    putData(groupEnrolmentKey, Json.toJson(groupEnrolment)) map (_ => true)
+
+  def subscriptionDetails(implicit request: Request[_]): Future[SubscriptionDetails] =
+    getData[SubscriptionDetails](subDetailsKey).map(_.getOrElse(SubscriptionDetails()))
+
+  def email(implicit request: Request[_]): Future[String] =
+    getData[String](emailKey).map(_.getOrElse(throwException(emailKey)))
 
-  def subscriptionDetails(implicit hc: HeaderCarrier): Future[SubscriptionDetails] =
-    getCached[SubscriptionDetails](sessionId, (cachedData, id) => cachedData.subscriptionDetails(id))
+  def eori(implicit request: Request[_]): Future[Option[String]] =
+    getData[String](eoriKey)
 
-  def email(implicit hc: HeaderCarrier): Future[String] =
-    getCached[String](sessionId, (cachedData, id) => cachedData.email(id))
+  def safeId(implicit request: Request[_]): Future[SafeId] = fetchSafeIdFromRegDetails.flatMap {
+    case Some(value) => Future.successful(value)
+    case None =>
+      fetchSafeIdFromReg06Response.map(
+        _.getOrElse(throw new IllegalStateException(s"$safeIdKey is not cached in data for the sessionId: $sessionId"))
+      )
+  }
 
-  def eori(implicit hc: HeaderCarrier): Future[Option[String]] =
-    getCached[Option[String]](sessionId, (cachedData, _) => cachedData.eori)
+  def fetchSafeIdFromReg06Response(implicit request: Request[_]): Future[Option[SafeId]] =
+    registerWithEoriAndIdResponse.map(
+      response =>
+        response.responseDetail.flatMap(_.responseData.map(_.SAFEID))
+          .map(SafeId(_))
+    ).recoverWith {
+      case _ => Future.successful(None)
```

Comment:
I would recommend changing this to filter out fatal JVM exceptions as we shouldn't catch those, e.g. we need to let the app crash `OutOfMemmoryError`.

Can be achieved using `case NonFatal(e) => ...`

More info: https://stackoverflow.com/a/66118268, https://stackoverflow.com/questions/29744462/the-difference-between-nonfatal-and-exception-in-scala